cong/tc.cc

```diff
@@ -611,41 +611,29 @@ namespace libsemigroups {
     partition_t classes;
 
     if (_cong._semigroup == nullptr) {
-      // Assert appropriate for JDM's long comment in cong.cc
-      assert(_cong._relations.empty() || _cong._extra.empty());
       if (_cong._extra.empty()) {
         return classes;  // trivial congruence - no nontrivial classes
       }
-      // nontrivial congruence on free semigroup - answer is infinite
       assert(!_cong._relations.empty());  // TODO: fail gracefully?
     }
 
     word_t word;
     // TODO: This may create many empty vectors that won't be used,
     //       because the classes are not numbered contiguously {0 .. n-1}
-    std::vector<std::vector<size_t>> pos_classes(_table.nr_rows(),
-                                                 std::vector<size_t>());
+    partition_t all_classes(_table.nr_rows(), class_t());
     // Look up each semigroup element's position in the table
     for (size_t pos = 0; pos < _cong._semigroup->size(); pos++) {
       _cong._semigroup->factorisation(word, pos);
       assert(word_to_class_index(word) < _table.nr_rows());
```

Comment:
At this point in TC we cannot rely on the indices being in a range `{0, ..., n - 1}` because we have not compressed the table.  Hence, we can only rely on it being between zero and the number of rows in the table.  The corresponding code in `kbfp.cc:100`, where we *can* rely on the range `{0, ..., n - 1}`, has an assert that checks the index is less than `nr_classes()`.